bergercookie/taskwarrior-syncall
# [draft] add tw asana

# Description

New, rudimentary sync service for TW<->Asana.

## Type of change

Please delete options that are not relevant.


# How Has This Been Tested?

With manual tests so far. Will add automated tests, but would like some initial high-level feedback.

## Setup

1. Create an Asana account
2. Generate an Asana Personal Access Token
3. Store Personal Access Token in password store
4. Find Asana workspace GID with `tw_asana_sync --list-asana-workspaces`

## Tests

Do these tests in-order.

- [x] Create task on Asana in workspace <workspace-gid>, sync to TW
   1. Create a task on Asana
   2. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is created in Asana with name of the Asana task set to the description of the TW task
   4. Verify and take note of a new combination is that save, it should look something like `tw_asana_sync --combination <workspace-gid>__None__test-asana`
- [x] Modify task on Asana sync to TW
   1. Modify description of last task, changing the name and due date
   2. Run the newly created combination `tw_asana_sync --combination <workspace-gid>__None__test-asana`
   3. Verify task is modified in TW, with updated `description` and a `due` fields
- [x] After no further changes, sync detects no changes
   1. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   2. Verify no changes are detected or applied to either Asana or TW
- [x] Modify task in TW sync to Asana
   1. In TW, Change `description` and remove `due` fields of last task 
   3. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   4. Verify task is modified in Asana, with updated name removed due date
- [x] After no further changes, sync detects no changes
   1. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   2. Verify no changes are detected or applied to either Asana or TW
- [x] Delete task in Asana, verify deleted in TW
   1. Delete the last task in Asana
   2. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify the task is marked as deleted (but not purged) in TW
- [x] Create untagged task in TW, is not synced to Asana
   1. Create a task in TW without the `test-asana` tag
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is not created in Asana
- [x] Add tag and due date to TW task, sync to Asana
   1. Add `test-asana` tag and a due date to last task
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is created in Asana with a due date
- [x] Remove TW due date, sync to Asana
   1. Remove due date from Asana task
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify due date is removed from Asana task
- [x] After no further changes, sync detects no changes
   1. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   2. Verify no changes are detected or applied to either Asana or TW
- [x] Complete TW task, sync to Asana
   1. Mark last task as completed in TW
   3.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   5. Verify task is marked completed in Asana
- [x] Un-complete task in Asana, sync to TW
   1. Mark last task as incomplete in Asana
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is marked incomlete in TW
- [x] Delete task in TW, sync to Asana
   1.  Delete task in TW
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is deleted (non-permanently) in Asana
    

# Checklist:

       ^ None were needed
